decode.c

```diff
@@ -284,6 +327,135 @@ static void mark_visible_input_output(struct solution *solution, struct input_ou
         mark_visible_region(solution, io->atoms[i].position, 0);
 }
 
+// TODO unhardcode this
+#define MAP_SIZE 32
+#define MAP_HALF_SIZE 16
+
+// maps u,v coordinates => cabinet id + 1 (0 means no cabinet)
+static uint8_t cabinet_map[MAP_SIZE][MAP_SIZE] = { 0 };
```

Comment:
Making this array global breaks threadsafety; could we have it be a local in `check_production_constraints` and pass it in to `cabinet_for_position`?